src/widgets/modals/ByProgramModal.js

```diff
@@ -0,0 +1,406 @@
+
+import { SETTINGS_KEYS } from '../../settings';
+import { getAllPrograms, getAllPeriodsForProgram } from '../../utilities/byProgram';
+import SequentialSteps from '../SequentialSteps';
+
+const localization = {
+  program: 'program',
+  supplier: 'supplier',
+  orderType: 'order type',
+  period: 'period',
+  programTitle: 'Select a Program to use',
+  supplierTitle: 'Select a Supplier to use',
+  orderTypeTitle: 'Select a Order Type to use',
+  periodTitle: 'Select a Period to use',
+  nameTitle: 'Give your stocktake a name',
+  programError: 'No programs available',
+  supplierError: 'No suppliers available',
+  orderTypeError: 'No order types available',
+  periodError: 'No periods available',
+};
+
+const getNewState = command => {
+  let newState;
+  switch (command) {
+    case 'RESET_ALL':
+      newState = {
+        program: {},
+        supplier: {},
+        period: {},
+        periods: null,
+        orderType: {},
+        orderTypes: null,
+        name: null,
+        currentStep: 0,
+      };
+      break;
+    case 'SELECT_PROGRAM':
+      newState = {
+        period: {},
+        periods: null,
+        supplier: {},
+        orderType: {},
+        orderTypes: null,
+        modalIsOpen: false,
+        currentStep: 1,
+      };
+      break;
+    case 'SELECT_ORDER_TYPE':
+      newState = { period: {}, periods: null, modalIsOpen: false };
+      break;
+    case 'SELECT_NAME':
+    case 'SELECT_PERIOD':
+    case 'SELECT_SUPPLIER':
+      newState = { modalIsOpen: false };
+      break;
+    default:
+      break;
+  }
+
+  return newState;
+};
+
+export class ByProgramModal extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      // Values used for deriving current UI state
+      currentStep: 0,
+      steps: null,
+      modalIsOpen: false,
+      isProgramBased: true,
+      // New requisition or stocktake values
+      program: {},
+      supplier: {},
+      orderType: {},
+      period: {},
+      name: null,
+      // Data stored for performance
+      storeTag: null,
+      programs: null,
+      suppliers: null,
+      orderTypes: null,
+      periods: null,
+    };
+  }
+
+  // Store unchanging select modal options and set the default
+  // steps array.
+  componentDidMount = () => {
+    const { settings, database } = this.props;
+    const programs = getAllPrograms(settings, database);
+    const suppliers = database.objects('Name').filtered('isSupplier = $0', true);
+    this.setCurrentSteps();
+    this.setState({ programs, suppliers });
+  };
+
+  getProgramValues = () => {
+    const { program, supplier, orderType, period, name } = this.state;
+    return { program, supplier, orderType, period, name };
+  };
+
+  getStepNumbers = () => {
+    const { isProgramBased } = this.state;
+    return {
+      program: 0,
+      supplier: isProgramBased ? 1 : 0,
+      orderType: 2,
+      period: 3,
+      name: isProgramBased ? 1 : 0,
```

Comment:
not a huge fan of keeping track of steps as number btw

Reply:
changed to use keys only 4ce5f8b35bc68f8ca5cd7a2d47e42bd383aa7e87